diffusion/models/layers.py

```diff
@@ -222,3 +223,208 @@ def __call__(
+        return x
+
+    def _reshape_from_attention(self, x: torch.Tensor, H: int, W: int) -> torch.Tensor:
+        """Reshape the input tensor from attention."""
+        # x is (B, H*W, C), need it to be (B, C, H, W) for conv
+        x = x.reshape(x.shape[0], H, W, -1).permute(0, 3, 1, 2).contiguous()
+        return x
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        """Forward through the attention layer."""
+        # Need to remember H, W to get back to it
+        H, W = x.shape[2:]
+        h = self.norm(x)
+        # Get q, k, and v
+        qkv = self.qkv_conv(h)
+        qkv = self._reshape_for_attention(qkv)
+        q, k, v = torch.split(qkv, self.input_channels, dim=2)
+        # Use torch's built in attention function
+        h = F.scaled_dot_product_attention(q, k, v, dropout_p=self.dropout_probability)
+        # Reshape back into an image style tensor
+        h = self._reshape_from_attention(h, H, W)
+        # Project to the output
+        h = self.proj_conv(h)
+        return h
+
+
+class Downsample(nn.Module):
+    """Downsampling layer that downsamples by a factor of 2.
+
+    Args:
+        input_channels (int): Number of input channels.
+        resample_with_conv (bool): Whether to use a conv for downsampling.
+    """
+
+    def __init__(self, input_channels: int, resample_with_conv: bool):
+        super().__init__()
+        self.input_channels = input_channels
+        self.resample_with_conv = resample_with_conv
+        if self.resample_with_conv:
+            self.conv = nn.Conv2d(self.input_channels, self.input_channels, kernel_size=3, stride=2, padding=0)
+            nn.init.kaiming_normal_(self.conv.weight, nonlinearity='linear')
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        if self.resample_with_conv:
+            # Need to do asymmetric padding to ensure the correct pixels are used in the downsampling conv
+            # and ensure the correct output size is generated for even sizes.
+            x = F.pad(x, (0, 1, 0, 1), mode='constant', value=0)
+            x = self.conv(x)
+        else:
+            x = F.avg_pool2d(x, kernel_size=2, stride=2)
+        return x
+
+
+class Upsample(nn.Module):
+    """Upsampling layer that upsamples by a factor of 2.
+
+    Args:
+        input_channels (int): Number of input channels.
+        resample_with_conv (bool): Whether to use a conv for upsampling.
+    """
+
+    def __init__(self, input_channels: int, resample_with_conv: bool):
+        super().__init__()
+        self.input_channels = input_channels
+        self.resample_with_conv = resample_with_conv
+        if self.resample_with_conv:
+            self.conv = nn.Conv2d(self.input_channels, self.input_channels, kernel_size=3, stride=1, padding=1)
+            nn.init.kaiming_normal_(self.conv.weight, nonlinearity='linear')
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x = F.interpolate(x, scale_factor=2, mode='nearest')
```

Comment:
Do we need tow rry about align corners or antialias? Probably not, but worth considering.

Reply:
Since it uses `mode='nearest'` `align_corners` doesn't have an effect. Antialiasing is probably a good idea, but I left it as default following [the original stable diffusion code](https://github.com/CompVis/stable-diffusion/blob/21f890f9da3cfbeaba8e2ac3c425ee9e998d5229/ldm/modules/diffusionmodules/model.py#L54).